CDAP-6894 Add a cache in NamespaceAdmin to cache NamespaceMeta

rohitsinha54 opened the pull request:
Issue: https://issues.cask.co/browse/CDAP-6894
Build: http://builds.cask.co/browse/CDAP-RUT20-1
ITN: http://builds.cask.co/browse/CDAP-ITM-37
- Added a LoadingCache in DefaultNamespaceQueryAdmin which is also used by the DefaultNamespaceAdmin.
- This is needed because we started supporting custom namespace which requires a look for NamespaceMeta through NamespaceQueryAdmin all over CDAP for different operations. This improves the performance by not making hbase call all the time. 
- The cache is loaded in get operation, refreshed in update and invalidated in delete.
- RemoteNamespaceQueryHandler now uses NamespaceAdmin rather than Store directly to take advantage of the cache. 

Note: 
- In the InMemoryNamespaceStore there is no cache since we don't really need one for unit tests.

anwar6953 commented:
There's only one instance of DefaultNamespaceAdmin bound?

rohitsinha54 commented:
Yes in AppFabricServiceRuntimeModule in Singleton.

anwar6953 commented on cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java:
```diff
@@ -124,25 +135,22 @@ public void apply(NamespaceMDS mds) throws Exception {
         }
       }, apps.get());
+    // refresh the namespacemeta in the cache
+    namespaceMetaCache.refresh(metadata.getNamespaceId());
```
why not simply invalidate it?

rohitsinha54 replied:
We can but I thought about that but decided to refresh  since the chance of having a get after update is high and then get will take less time which is preferable. Either is fine I guess.

anwar6953 replied:
fair enough. both seem fine to me.

anwar6953 commented:
There's a binding to `DefaultNamespaceStore` in ExploreServiceTwillRunnable. That can have issues, if that caches separately, right?

bdmogal commented:
@rohitsinha54 please add some tests for the cache as well. should make sure that the cache is hit and missed at appropriate times.

chtyim commented on cdap-data-fabric/src/main/java/co/cask/cdap/store/DefaultNamespaceStore.java:
```diff
@@ -90,6 +94,13 @@ public TransactionExecutor get() {
       }
     };
+    this.namespaceMetaCache = CacheBuilder.newBuilder().build(new CacheLoader<NamespaceId, NamespaceMeta>() {
+      @SuppressWarnings("NullableProblems")
```
No need to have this.

anwar6953 commented:
@bdmogal any suggestions on how to validate that, without inserting hooks into the methods of the DefaultNamespaceStore?

bdmogal commented:
expose the cache as a map via a @VisibleForTesting method and check its state after various calls to DefaultNamespaceStore?

anwar6953 commented:
@rohitsinha54 We're going to move the cache out of the store, right?

rohitsinha54 commented:
@anwar6953 : Yes have to see if we can move it into the Admin. I will get to this PR in sometime today. Working on other issue currently.

Commit:
CDAP-6894 Cache NamespaceMeta

Move DefaultNamespaceQueryAdmin into DefaultNamespaceAdmin

Fix circular dependency

- Together with some other fixes.

Addressed comments

Fix RemoteNamespaceQueryTest

Fix namespace deletion logic.

Resolves circular dependency.

Make RemoteNamespaceQueryClient use AppFabric

Addressed comments

Addressed comments

Addressed comments

Fix unit tests

WIP

Addressed comments and fixes

Use master module

Add StorageProviderNamespaceAdmin binding in PreviewServerModule

Fix SimpleNamespaceQueryAdmin for unit tests

anwar6953 commented:
Can you update the description for the design change? Most people read the description, as a summary of the changes, and its inaccurate right now (since the cache has been moved out of the NamespaceStore)

rohitsinha54 commented:
@anwar6953 : I have updated the description.

anwar6953 commented on cdap-app-fabric/src/main/java/co/cask/cdap/gateway/handlers/meta/RemoteNamespaceQueryHandler.java:
```diff
@@ -35,24 +35,24 @@
 @Path(AbstractRemoteSystemOpsHandler.VERSION + "/execute")
 public class RemoteNamespaceQueryHandler extends AbstractRemoteSystemOpsHandler {
```
I think we should remove this handler, and change the client to use the handler in master instead.

chtyim replied:
The name is not good. Remote means it is talking to remote, but it seems like this one is actually not? Or a bit unclear.

anwar6953 replied:
Yea, the naming is not good.
I think we should remove this handler now.
@bdmogal @gokulavasan ?

rohitsinha54 replied:
@bdmogal @gokulavasan : Any thoughts ?

bdmogal replied:
yes, we could just use APIs from NamespaceHttpHandler?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -406,6 +407,8 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
 
     nsStore.update(builder.build());
+    // refresh the cache with new meta
+    namespaceMetaCache.refresh(namespaceId.toEntityId());
```
Will `DefaultNamespaceAdmin` and the `DefaultNamepsaceQueryAdmin` binding creating different instances?

rohitsinha54 replied:
@chtyim I have this in AppFabricServiceRuntimeModule

```
bind(NamespaceAdmin.class).to(DefaultNamespaceAdmin.class).in(Scopes.SINGLETON);    
bind(NamespaceQueryAdmin.class).to(DefaultNamespaceQueryAdmin.class).in(Scopes.SINGLETON);
```

Since NamespaceQueryAdmin is being binded in Singleton scope NamespaceAdmin which extends NamesapceQueryAdmin will use the same instance. Please correct me, if my understanding is wrong.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceQueryAdmin.java:
```diff
@@ -83,16 +97,18 @@ public boolean apply(NamespaceMeta namespaceMeta) {
+      return namespaceMetaCache.get(namespaceId.toEntityId());
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof NamespaceNotFoundException) {
+        throw new NamespaceNotFoundException(namespaceId);
```
Why create a new one? It'll lose the stacktrace. Should you just throw `e.getCause()` in this catch block?

rohitsinha54 replied:
e.getCause() will give a throwable so to throw that we will need to change the API to throw Throwables which will need all caller to handle throwables. We can wrap it in a Exception but I thought wrapping specifically will be better idea.

anwar6953 replied:
you can caste it to a `NamespaceNotFoundException` and then throw it.

rohitsinha54 commented:
New
Build: http://builds.cask.co/browse/CDAP-RUT20-4
ITM: http://builds.cask.co/browse/CDAP-ITM0-19

rohitsinha54 commented:
Combining the implementation of NamespaceQueryAdmin and NamespaceAdmin in one class DefaultNamespaceAdmin is causing some circular dependency issue in guice bindings. Working on solving those.

bdmogal commented:
@rohitsinha54 didn't we make `DefaultNamespaceQueryAdmin` a separate class to avoid having to inject all the dependencies that `DefaultNamespaceAdmin` needs every time? How come we don't need that separation anymore?

rohitsinha54 commented:
@bdmogal Yes, we decided to merge them together but keep them as separate interface. The idea was that since Guice will create proxy and handle it. But it seems like Guice is proxying for NamespaceQueryAdmin/NamespaceAdmin but there are thing which are needed by DefaultNamespaceAdmin which is causing their own circular dependencies and since some of them are not interface guice fails to  proxy these.
I am going to revert the change of merging them in one class. @chtyim Please let me know, if you disagree and if so, please suggest some other solution to this.

rohitsinha54 commented:
The first three commits are part of another PR: https://github.com/caskdata/cdap/pull/6503
Move Impersonator and related classes to cdap-common
Remove the not so useful ImpersonationUserResolver
Make Impersonator an interface

anwar6953 commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -42,26 +47,38 @@
+        return new RandomEndpointStrategy(discoveryClient.discover(Constants.Service.APP_FABRIC_HTTP));
       }
     });
+    String masterPrincipal = cConf.get(Constants.Security.CFG_CDAP_MASTER_KRB_PRINCIPAL);
```
put this in the `if (SecurityUtil.isKerberosEnabled(cConf))` block

anwar6953 commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -70,6 +87,17 @@ private InetSocketAddress getNamespaceServiceAddress() {
+    if (masterShortUserName != null) {
+      username = masterShortUserName;
+    } else {
+      username = UserGroupInformation.getCurrentUser().getShortUserName();
```
just put this in the constructor itself.
`username` can be cached. It doesn't need to be recomputed each time.
Also, probably rename it something more specific. `username` is too ambiguous.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServerModule.java:
```diff
@@ -81,7 +81,7 @@ protected void configure() {
 
     bind(NamespaceQueryAdmin.class).to(DefaultNamespaceAdmin.class).in(Scopes.SINGLETON);
-    expose(NamespaceQueryAdmin.class);
+    expose(DefaultNamespaceAdmin.class);
```
Why exposes the implementation? Usually we should only the interface.

rohitsinha54 replied:
Oooops! will change. I guess did it during resolving conflicts with my brain switched off.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -390,11 +347,91 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
+      Iterables.filter(namespaces, new com.google.common.base.Predicate<NamespaceMeta>() {
+        @Override
+        public boolean apply(NamespaceMeta namespaceMeta) {
+          return filter.apply(new NamespaceId(namespaceMeta.getName()));
```
`namespaceMeta.getNamespaceId()`.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -390,11 +347,91 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
+      return namespaceMetaCache.get(namespaceId.toEntityId());
+    } catch (ExecutionException e) {
+      if (e.getCause() instanceof NamespaceNotFoundException) {
+        throw (NamespaceNotFoundException) e.getCause();
```
No need to cast.

chtyim replied:
Oh.. you'll need to cast to Exception. Actually you can combine the ifs into one

``` java
Throwable cause = e.getCause();
if (cause instanceof NamespaceNotFoundException || cause instanceof IOException || cause instanceof UnauthorizedException) {
  throw (Exception) cause;
}
throw e;
```

anwar6953 commented:
I don't have any additional comments; lgtm.
Just make sure it works on auth-enabled cluster

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -390,11 +347,89 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
+    if (ns == null) {
+      throw new NamespaceNotFoundException(namespaceId.toId());
+    }
+    Principal principal = authenticationContext.getPrincipal();
```
please move this out of the cache and into the `get` method. Also, `exists()` should not enforce authorization

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -390,11 +347,89 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
+      throw new NamespaceNotFoundException(namespaceId.toId());
+    }
+    Principal principal = authenticationContext.getPrincipal();
+    co.cask.cdap.api.Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
```
CDAP class should not be fully qualified

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -390,11 +347,89 @@ public synchronized void updateProperties(Id.Namespace namespaceId, NamespaceMet
+  public List<NamespaceMeta> list() throws Exception {
+    List<NamespaceMeta> namespaces = nsStore.list();
+    Principal principal = authenticationContext.getPrincipal();
+    final co.cask.cdap.api.Predicate<EntityId> filter = authorizationEnforcer.createFilter(principal);
```
CDAP class should not be fully qualified

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceResourceDeleter.java:
```diff
@@ -0,0 +1,134 @@
+
+  private void deleteMetrics(NamespaceId namespaceId) throws Exception {
+    long endTs = System.currentTimeMillis() / 1000;
+    Map<String, String> tags = Maps.newHashMap();
```
`new HashMap<>()`

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceExistenceVerifier.java:
```diff
@@ -19,28 +19,37 @@
-      if (nsStore.get(namespaceId.toId()) == null) {
-        throw new NamespaceNotFoundException(namespaceId.toId());
+      try {
+        if (namespaceQueryAdmin.get(namespaceId.toId()) == null) {
```
`get()` already throws a `NamespaceNotFoundException` below right? Why do we need this?

rohitsinha54 replied:
I just wrapped the code didn't notice the condition. Get was never returning null. The API is not even marked Nullable.

rohitsinha54 replied:
@bdmogal I am changing this to call exists rather than get which has authorization. ensuring SYSTEM Namespace should not need authorization. Let me know if you have a different opinion.

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -70,6 +89,10 @@ private InetSocketAddress getNamespaceServiceAddress() {
+  }
+
+  private HttpRequest addUserIdHeader(HttpRequest request) throws IOException {
+    return new HttpRequest.Builder(request).addHeader(Constants.Security.Headers.USER_ID, userId).build();
```
can the `userId` be null? What happens when it is null? I think it should be ok, but good to test, otherwise add a guard.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceExistenceVerifier.java:
```diff
@@ -41,9 +41,7 @@
-        if (namespaceQueryAdmin.get(namespaceId.toId()) == null) {
-          throw new NamespaceNotFoundException(namespaceId.toId());
-        }
+        namespaceQueryAdmin.exists(namespaceId.toId());
```
nit: there's no namespaceQueryAdmin now, right? rename to namespaceAdmin?

rohitsinha54 replied:
We do have NamespaceQueryAdmin the interface we just moved the implementation in DefaultNamespaceAdmin. Since here we are just querying we are  working with QueryAdmin interface.

bdmogal replied:
ok. the contract of this method though is that it should throw a NamespaceNotFoundException if the namespace does not exist. I don't think that is obeyed here.

bdmogal commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceExistenceVerifier.java:
```diff
@@ -41,9 +41,7 @@
-        }
+        namespaceQueryAdmin.exists(namespaceId.toId());
       } catch (Exception e) {
         if (e instanceof NamespaceNotFoundException) {
```
can this ever happen? otherwise just remove it?

bdmogal commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -66,6 +67,7 @@ public EndpointStrategy get() {
         // if kerberos is not enabled then fall back to UGI to get username
         this.userId = UserGroupInformation.getCurrentUser().getShortUserName();
       }
+      Preconditions.checkNotNull(userId, "Failed to obtain username.");
```
please test to make sure that this is ok.

rohitsinha54 replied:
Can you suggest  how can we make UGI getCurrentUser return null ?

bdmogal commented:
couple minor comments on the last commit

bdmogal commented:
1 comment about a comment, then LGTM

rohitsinha54 commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -42,26 +48,41 @@
       }
     });
+    try {
+      if (SecurityUtil.isKerberosEnabled(cConf)) {
```
This does not seem to work for impersonated namespace. The issue is that we  cannot always use cdap master keytab since cdap user will have no permission on an impersonated namespace being created by the user. The namespace meta for the namespace will belong to the user and any get operation on this namespace will fail. 
I think we should pass the AuthenticationContext.getPrincipal().getName() if kerberos is enabled.
@bdmogal : Any suggestions on how to solve this ?

anwar6953 replied:
yea, see if you can use SecurityRequestContext, if its set?

bdmogal replied:
@rohitsinha54 two things about BaseHiveExploreService#getHiveDatabase:
1. It should have a version that takes in a `NamespaceMeta`, instead of just the namespace name. This will be called from `create`, where you already have the `NamespaceMeta` before you call `getHiveDb`. So in this case you don't need to make a call to `namespaceQueryAdmin.get()`.
2. For cases other than `create`, just use the userId from `AuthenticationContext#getPrincipal()`. 

This needs to be tested thoroughly though.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/app/preview/PreviewServerModule.java:
```diff
@@ -80,7 +80,7 @@ protected void configure() {
     expose(RuntimeStore.class);
 
-    bind(NamespaceQueryAdmin.class).to(DefaultNamespaceQueryAdmin.class).in(Scopes.SINGLETON);
+    bind(NamespaceQueryAdmin.class).to(DefaultNamespaceAdmin.class).in(Scopes.SINGLETON);
```
Don't you need the same bindings as in `AppFabricServiceRuntimeModule`? Otherwise the DNA will miss the resource deleter.

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/DefaultNamespaceAdmin.java:
```diff
@@ -272,69 +266,31 @@ public synchronized void delete(final Id.Namespace namespaceId) throws Exception
+        // so no one can clean up. This may result in orphaned privileges, which will be cleaned up by the create API
+        // if the same namespace is successfully re-created.
+        privilegesManager.revoke(namespace);
+        LOG.info("Namespace '{}' deleted", namespace);
```
Isn't that you need to invalid the cache?

chtyim replied:
I see it is done in the `deleteNamespaceMeta`. All good.

chtyim commented on cdap-common/src/main/java/co/cask/cdap/common/namespace/RemoteNamespaceQueryClient.java:
```diff
@@ -42,26 +48,41 @@
+    try {
+      if (SecurityUtil.isKerberosEnabled(cConf)) {
+        // if kerberos is enabled use the name from cdap master principal
+        String masterPrincipal = cConf.get(Constants.Security.CFG_CDAP_MASTER_KRB_PRINCIPAL);
```
What is this for? This is a remote client, which means it should haven't Kerberos login. Should it just like any other http client that set the USER_ID in the header?

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceExistenceVerifier.java:
```diff
@@ -19,27 +19,35 @@
+      try {
+        exists = namespaceQueryAdmin.exists(namespaceId.toId());
+      } catch (Exception e) {
+        Throwables.propagate(e);
```
`throw Throwables.propagate`

chtyim commented on cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/namespace/NamespaceExistenceVerifier.java:
```diff
@@ -19,27 +19,35 @@
+      } catch (Exception e) {
+        Throwables.propagate(e);
+      }
+      if (!exists) {
```
Move this inside the `try` block. Simply do

``` java
if (!namespaceQueryAdmin.exists(namespaceId.toId())) {
  throw ...
}
```

rohitsinha54 replied:
According to the contract of  this method if namespace does not exists we throw NotFoundExeption. How will that work ? since exists returns false if namespace does not exists.

rohitsinha54 commented:
Running final build here: http://builds.cask.co/browse/CDAP-RUT20-9

chtyim commented:
LGTM. Please squash to one commit before merge.
